src/pyats/contrib/creators/topology.py

```diff
@@ -0,0 +1,656 @@
+            return 'nxos'
+
+    def get_interfaces_ip_address(self, device):
+        '''
+        get the ip address for all of the generated interfaces
+        on the give device
+        '''
+        if (not device.is_connected() or device.os not in os_list 
+            or len(device.interfaces) < 1):
+            return
+        for interface in device.interfaces.values():
+            if interface.ipv4 is None or interface.ipv6 is None:
+                ip = device.api.get_interface_ip_address(interface.name)
+                if ip:
+                    ip = ipaddress.IPv4Interface(ip)
+                    interface.ipv4 = ip
+        
+
+    def create_yaml_dict(self, testbed, f1):
+        '''
+        Take the information laid out in the testbed and then format it into a
+        dictionary to be integrated with the existing
+        '''
+        log.info('Creating dictionary based on testbed')
+        yaml_dict = {'devices':{}, 'topology': {}}
+        credential_dict, _ = self.get_credentials_and_proxies(f1)
+
+        # write new devices into dict
+        for device in testbed.devices.values():
+            log.info('Adding device info for {}'.format(device.name))
+            if device.name not in f1['devices']:
+                yaml_dict['devices'][device.name] = {'type': device.type,
+                                                    'os': device.os,
+                                                    'credentials': credential_dict,
+                                                    'connections': {}
+                                                    }
+                c = yaml_dict['devices'][device.name]['connections']
+                for connect in device.connections:
+                    try:
+                        c[connect] = {'protocol':device.connections[connect]['protocol'],
+                                      'ip': device.connections[connect]['ip'],
+                                    }
+                    except:
+                        continue
+
+            # write in the interfaces and link from devices into testbed
+            interface_dict = {'interfaces': {}}
+            log.info('Adding interface infor for {}'.format(device.name))        
+            for interface in device.interfaces.values():
+                interface_dict['interfaces'][interface.name] = {'type': interface.type}
+                if interface.link is not None:
+                    interface_dict['interfaces'][interface.name]['link'] = interface.link.name
+                if interface.ipv4 is not None:
+                    interface_dict['interfaces'][interface.name]['ipv4'] = str(interface.ipv4)
+            # add interface information into the topology part of yaml_dict
+            yaml_dict['topology'][device.name] = interface_dict
+        log.info('topology discovered is: {}'.format(yaml_dict['topology']))
+        return yaml_dict
+
+    def _generate(self):

```

Comment:
this is so big, can it also be broken down in smaller functions and just call them

Reply:
Done